doc/tutorials/rtvm_performance_tips.rst

```diff
@@ -0,0 +1,213 @@
+by the vCPU.
+
+ACRN eliminates almost all VM-exits triggered by external events via the
+LAPIC passthrough. A few exceptions exist:
+
+- SMI - it will bring the processor into the SMM, causing a much longer
+  performance impact. The SMI should be handled in the BIOS.
+
+- NMI - ACRN uses NMI for system-level notification.
+
+Users should take care of VM-exits that are triggered by operations
+initiated by the vCPU. Refer to the Intel SMD: "Instructions Cause VM-exits
+Unconditionally" (SDM V3, 25.1.2) and "Instructions That Cause VM-exits
+Conditionally" (SDM V3, 25.1.3).
+
+**Tip 3:** Do not use CPUID in the RT critical section.
+
+CPUID is an instruction that causes VM-exits unconditionally. As to the
+normal usage of CPUID, this can be avoided by detecting the CPU capability
+before entering the RT critical section. CPUID can be executed at any
+privilege level to serialize instruction execution and its high efficiency
+of execution. It's commonly used as a serializing instruction in an
+application, and a typical case is using CPUID immediately before and after
+RDTSC. In order to remove CPUID in this case, use RDTSCP instead of RDTSC.
+Because RDTSCP waits until all previous instructions have been executed
+before reading the counter, and the subsequent instructions after the RDTSCP
+normally have data dependency on it, they must wait until the RDTSCP has
+been executed.
+
+RDMSR or WRMSR are instructions that cause VM-exits conditionally. On the
+ACRN RTVM, most MSRs are not intercepted by the HV, so they won't cause a
+VM-exit. But there are exceptions for security consideration: 1) read from
+APICID and LDR; 2) write to TSC_ADJUST if VMX_TSC_OFFSET_FULL is zero;
+otherwise, read and write to TSC_ADJUST and TSC_DEADLINE; 3) write to ICR.
+
+**Tip 4:** Do not use RDMSR to access APICID and LDR at the RT critical
+section.
+
+ACRN does not intend to present a physical APICID to a guest so that APICID
+and LDR are virtualized even though LAPIC is passthrough. As a result,
+access to APICID and LDR can cause a VM-exit.
+
+**Tip 5:** Guarantee that VMX_TSC_OFFSET_FULL is zero; otherwise, do not
+access TSC_ADJUST and TSC_DEADLINE in the RT critical section.
+
+ACRN uses VMX_TSC_OFFSET_FULL as the offset between vTSC_ADJUST and
+pTSC_ADJUST; therefore, if VMX_TSC_OFFSET_FULL is zero, intercepting
+TSC_ADJUST and TSC_DEADLINE is not necessary. Otherwise, they should be
+intercepted to guarantee functionality.
+
+**Tip 6:** Utilize Preempt-RT Linux mechanisms to reduce the access of ICR
+from the RT core:
+
+#. Add "domain" to the "isolcpus" ( “isolcpus=nohz,domain,1” ) to the kernel parameters.
+#. Add "idle=poll" to the kernel parameters.
+#. Add "rcu_nocb_poll" along with "rcu_nocbs=1" to the kernel parameters.
+#. Disable the logging service like journald, syslogd if possible.
+
+These parameters are recommended for the guest Preempt-RT Linux. For a UP
+RTVM, ICR interception is not a problem. But for an SMP RTVM, IPI may be

```

Comment:
Can we spell out IP once (Intel-Process Interrupt) and ICR (Interrupt Command Register)